### PR TITLE
Add scroll horizontally support for paragraph

### PR DESCRIPTION
**Pull request by xiaopengli89**
Now Paragraph can scroll horizontally
```rust
let p = Paragraph::new(items)
    .block(
        Block::default()
    )
    .scroll((0, 7));
```

```
+------------------------------+
|hello world                   |
+------------------------------+
```

```
+------------------------------+
|orld                          |
+------------------------------+
```

### CODE REVIEW

**Comment by fdehau** [comment]
Thanks that looks like a nice addition. Could you add some tests to cover this new behavior ?

**Comment by xiaopengli89** [comment]
I've add some tests

File: src/widgets/paragraph.rs [hunk 1]
```diff
@@ -136,21 +136,28 @@ where
...
         let mut y = 0;
         while let Some((current_line, current_line_width)) = line_composer.next_line() {
-            if y >= self.scroll {
+            if y >= self.scroll.0 {
                 let mut x = get_line_offset(current_line_width, text_area.width, self.alignment);
                 for Styled(symbol, style) in current_line {
-                    buf.get_mut(text_area.left() + x, text_area.top() + y - self.scroll)
-                        .set_symbol(symbol)
+                    buf.get_mut(text_area.left() + x, text_area.top() + y - self.scroll.0)
+                        .set_symbol(if symbol.is_empty() { " " } else { symbol })
```
Comment (fdehau) [inline src/widgets/paragraph.rs:154]: Why was that change required ?

File: src/widgets/reflow.rs [hunk 2]
```diff
@@ -124,16 +125,19 @@ pub struct LineTruncator<'a, 'b> {
     symbols: &'b mut dyn Iterator<Item = Styled<'a>>,
     max_line_width: u16,
     current_line: Vec<Styled<'a>>,
+    cut: u16,
```
Comment (fdehau) [inline src/widgets/reflow.rs:128]: I'm not a big fan of the name. What do think of `horizontal_offset` ? Could you also add a doc comment to explain how is this property will be used when the `LineTrucator` process lines ?

File: src/widgets/reflow.rs [hunk 3]
```diff
@@ -124,16 +125,19 @@ pub struct LineTruncator<'a, 'b> {
...
 }
 
 impl<'a, 'b> LineTruncator<'a, 'b> {
     pub fn new(
         symbols: &'b mut dyn Iterator<Item = Styled<'a>>,
         max_line_width: u16,
+        cut: u16,
```
Comment (fdehau) [inline src/widgets/reflow.rs:135]: Given that it is not a required property, could we default to `0` and have a different method to set it ?

File: src/widgets/reflow.rs [hunk 4]
```diff
@@ -189,6 +207,20 @@ impl<'a, 'b> LineComposer<'a> for LineTruncator<'a, 'b> {
     }
 }
 
+fn trim_width(src: &str, mut wid: usize) -> &str {
```
Comment (fdehau) [inline src/widgets/reflow.rs:210]: Even though the function name is quite explicit, could you add a small doc comment on what this function does ?

File: src/widgets/paragraph.rs (hunk 1 above)
Comment (xiaopengli89) [reply src/widgets/paragraph.rs:154]: Thanks for your reviews. If the symbol is empty, the last char which rendered last time will leave on the line. May be there are other better solutions.

File: src/widgets/reflow.rs (hunk 2 above)
Comment (xiaopengli89) [reply src/widgets/reflow.rs:128]: I agree with you, and will add a doc to explain.

File: src/widgets/reflow.rs (hunk 3 above)
Comment (xiaopengli89) [reply src/widgets/reflow.rs:135]: That's a good idea.

File: src/widgets/reflow.rs (hunk 4 above)
Comment (xiaopengli89) [reply src/widgets/reflow.rs:210]: Okay.

File: src/widgets/paragraph.rs (hunk 1 above)
Comment (fdehau) [reply src/widgets/paragraph.rs:151]: No that seems right but could you add a comment to explain this. It will make the code easier to follow next time somebody has to read this part of the code :).

Comment (xiaopengli89) [reply src/widgets/paragraph.rs:151]: Agree with you.

**Comment by fdehau** [comment]
Thanks a lot ! That's a nice addition.
